### DISCUSSION TITLE
When search param is set to a stringified number ("11000"), validating or getting the param returns a number but is typed as string

### DISCUSSION THREAD

**Discussion by maxwellwalin** [TanStack/router#430] [Q&A]
I am currently navigating to a URL with the following search:
```
navigate({
  search(current) {
    return {
      ...current,
      clientId: "11000",
    };
  },
});
```
The clientId can be any string, including a stringified number like the above. The problem is that when parsing or getting this search parameter, it evaluates as a number. 

I've attempted using preSearchFilters and postSearchFilters as I do not see a method of parsing/stringifying the search available, but that did not work (as I assume that is not what those options are for)!

Please let me know if this is a bug, or I am blindly missing the option that would resolve this. I appreciate the help.

**Comment by briavicenti** [comment]
Hey @maxwellwalin I know it's been a while, but I call out a (limited) solution for this issue in my thread [here](https://github.com/TanStack/router/discussions/553)! I'm hoping for the ability to have more nuanced parsing behavior per usage of `useSearch`.

**Comment by rahulvils** [comment]
@maxwellwalin @briavicenti  For now you can use a preprocess using Zod
```tsx
import { createFileRoute } from '@tanstack/react-router';
import { z } from 'zod';

const resetPasswordSchema = z.object({
  token: z.preprocess((val) => (typeof val === 'number' ? val.toString() : val), z.string().optional()),
});

export const Route = createFileRoute('/auth/reset-password/')({
  validateSearch: (search) => resetPasswordSchema.parse(search),
});
```

**Comment by andybarron** [comment]
i strongly believe the default behavior is incorrect. search params are (for better or worse) strings, and auto-magically converting them to numbers or booleans is very error-prone. especially since tanstack-router offers a really nice API for typing your query string via `validateSearch`, pre-processing them before that is misleading.

in particular, if i specify that a parameter is an optional string, the current behavior will fail validation if the users tries to search for any string that could be parsed as a number. i really don't think this is a sensible default.

**Reply by krambono** [reply]
It looks like we can change the default parsing/stringify at the router level.

I just tried this, and it seems to work!

```ts
const router = createRouter({
  routeTree,
  stringifySearch: search => {
    const params = new URLSearchParams(search).toString();
    if (params.trim()) {
      return `?${params}`;
    }
    return '';
  },
  parseSearch: search => Object.fromEntries(new URLSearchParams(search)),
});
```
